### PR TITLE
[Core] Add placeholder callback methods as members of RayDaskCallback

### PR DESCRIPTION
**Pull request by peytondmurray**
## Why are these changes needed?

This PR adds the overridable methods for `RayDaskCallback` to the base class.

Previously, the user had a few choices for instantiating a `RayDaskCallback`:

```python
RayDaskCallback(ray_presubmit=my_custom_presubmit_func)
```

or

```python
with RayDaskCallback(ray_presubmit=my_custom_presubmit_func) as cb:
    ...
```

or by subclassing:

```python
class MyCallback(RayDaskCallback):

    def _ray_presubmit(self, task, key, deps):
        ....
```

Currently the constructor for `RayDaskCallback` dynamically generates the callback methods and attaches them to the class instance upon instantiation using `setattr`. This pattern doesn't work with developer tooling including language servers, static code analysis tools, or documentation generators because these functions are generated dynamically. Additionally the documentation for dask-on-ray instructs users to use the third approach (using subclassing) rather than passing methods to the constructor. Finally, the docstring for `RayDaskCallback.__init__` currently has function signatures and pseudo-docstrings for the callback methods the user should be implementing. This docstring is itself malformed, so Sphinx mistakenly tries to interpret parts of this as references which do not exist.

At the request of @c21 I've actually implemented these functions as they appear in the docstring. Type hints are incomplete here, but match what was already in the previous docstring. I also added a summary in the dask-on-ray documentation.

## Related issue number

Unblocks #39658.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by peytondmurray** [comment]
Tagging @c21 for code review, @angelinalg for docs changes.

File: python/ray/util/dask/callbacks.py [hunk 1]
```diff
@@ -156,6 +79,99 @@ def unregister(self):
         type(self).ray_active.remove(self._ray_callback)
         super().unregister()
 
+    def _ray_presubmit(self, task, key, deps) -> Optional[Any]:
+        """Run before submitting a Ray task.
+
+        If this callback returns a non-`None` value, a Ray task will _not_ be
```
Comment (angelinalg) [inline python/ray/util/dask/callbacks.py:85]:
```suggestion
        If this callback returns a non-`None` value, Ray does _not_ create 
```

File: python/ray/util/dask/callbacks.py [hunk 2]
```diff
@@ -156,6 +79,99 @@ def unregister(self):
...
+        created and this value will be used as the would-be task's result value.
```
Comment (angelinalg) [inline python/ray/util/dask/callbacks.py:86]:
```suggestion
        a task and uses this value as the would-be task's result value.
```

File: python/ray/util/dask/callbacks.py [hunk 3]
```diff
@@ -156,6 +79,99 @@ def unregister(self):
...
+
+        Args:
+            task: A Dask task, where the first tuple item is
+                the task function, and the remaining tuple items are
+                the task arguments (either the actual argument values,
```
Comment (angelinalg) [inline python/ray/util/dask/callbacks.py:91]:
```suggestion
                the task arguments, which are either the actual argument values,
```

File: python/ray/util/dask/callbacks.py [hunk 4]
```diff
@@ -156,6 +79,99 @@ def unregister(self):
...
+                or Dask keys into the deps dictionary whose
+                corresponding values are the argument values).
```
Comment (angelinalg) [inline python/ray/util/dask/callbacks.py:93]:
```suggestion
                corresponding values are the argument values.
```

File: python/ray/util/dask/callbacks.py [hunk 5]
```diff
@@ -156,6 +79,99 @@ def unregister(self):
...
+            key: The Dask graph key for the given task.
+            deps: The dependencies of this task.
+
+        Returns:
+            Either None, in which case a Ray task will be submitted, or
```
Comment (angelinalg) [inline python/ray/util/dask/callbacks.py:98]:
```suggestion
            Either None, in which case Ray submits a task, or
```

File: python/ray/util/dask/callbacks.py [hunk 6]
```diff
@@ -156,6 +79,99 @@ def unregister(self):
...
+            a non-None value, in which case a Ray task will not be
```
Comment (angelinalg) [inline python/ray/util/dask/callbacks.py:99]:
```suggestion
            a non-None value, in which case Ray task doesn't submit
```

File: python/ray/util/dask/callbacks.py [hunk 7]
```diff
@@ -156,6 +79,99 @@ def unregister(self):
...
+            submitted and this return value will be used as the
```
Comment (angelinalg) [inline python/ray/util/dask/callbacks.py:100]:
```suggestion
            a task and uses this return value as the
```

File: python/ray/util/dask/callbacks.py [hunk 8]
```diff
@@ -156,6 +79,99 @@ def unregister(self):
...
+        """
+        pass
+
+    def _ray_postsubmit(self, task, key, deps, object_ref: ObjectRef):
+        """Run after submitting a Ray task.
+
+        Args:
+            task: A Dask task, where the first tuple item is
+                the task function, and the remaining tuple items are
+                the task arguments (either the actual argument values,
```
Comment (angelinalg) [inline python/ray/util/dask/callbacks.py:111]:
```suggestion
                the task arguments, which are either the actual argument values,
```

File: python/ray/util/dask/callbacks.py [hunk 9]
```diff
@@ -156,6 +79,99 @@ def unregister(self):
...
+                or Dask keys into the deps dictionary whose
+                corresponding values are the argument values).
```
Comment (angelinalg) [inline python/ray/util/dask/callbacks.py:113]:
```suggestion
                corresponding values are the argument values.
```

File: python/ray/util/dask/callbacks.py [hunk 10]
```diff
@@ -156,6 +79,99 @@ def unregister(self):
...
+            object_ref (ray.ObjectRef): The object reference for the
+                return value of the Ray task.
+
+        """
+        pass
+
+    def _ray_pretask(self, key, object_refs: List[ObjectRef]):
+        """Run before executing a Dask task within a Ray task.
+
+        This executes after the task has been submitted, within a Ray
```
Comment (angelinalg) [inline python/ray/util/dask/callbacks.py:125]:
```suggestion
        This method executes after Ray submits the task within a Ray
```

File: python/ray/util/dask/callbacks.py [hunk 11]
```diff
@@ -156,6 +79,99 @@ def unregister(self):
...
+        worker. The return value of this task will be passed to the
```
Comment (angelinalg) [inline python/ray/util/dask/callbacks.py:126]:
```suggestion
        worker. Ray passes the return value of this task to the
```

File: python/ray/util/dask/callbacks.py [hunk 12]
```diff
@@ -156,6 +79,99 @@ def unregister(self):
...
+        _ray_posttask callback, if provided.
+
+        Args:
+            key: The Dask graph key for the Dask task.
+            object_refs (List[ray.ObjectRef]): The object references
+                for the arguments of the Ray task.
+
+        Returns:
+            A value that will be passed to the corresponding
```
Comment (angelinalg) [inline python/ray/util/dask/callbacks.py:135]:
```suggestion
            A value that Ray passes to the corresponding
```

File: python/ray/util/dask/callbacks.py [hunk 13]
```diff
@@ -156,6 +79,99 @@ def unregister(self):
...
+            _ray_posttask callback, if said callback is defined.
```
Comment (angelinalg) [inline python/ray/util/dask/callbacks.py:136]:
```suggestion
            _ray_posttask callback, if the callback is defined.
```

File: python/ray/util/dask/callbacks.py [hunk 14]
```diff
@@ -156,6 +79,99 @@ def unregister(self):
...
+        """
+        pass
+
+    def _ray_posttask(self, key, result, pre_state):
+        """Run after executing a Dask task within a Ray task.
+
+        This executes within a Ray worker. This callback receives the
```
Comment (angelinalg) [inline python/ray/util/dask/callbacks.py:143]:
```suggestion
        This method executes within a Ray worker. This callback receives the
```

File: python/ray/util/dask/callbacks.py [hunk 15]
```diff
@@ -156,6 +79,99 @@ def unregister(self):
...
+        Args:
+            key: The Dask graph key for the Dask task.
+            result: The task result value.
+            pre_state: The return value of the corresponding
+                _ray_pretask callback, if said callback is defined.
+        """
+        pass
+
+    def _ray_postsubmit_all(self, object_refs: List[ObjectRef], dsk):
+        """Run after all Ray tasks have been submitted.
```
Comment (angelinalg) [inline python/ray/util/dask/callbacks.py:155]:
```suggestion
        """Run after Ray submits all tasks.
```

File: python/ray/util/dask/callbacks.py [hunk 16]
```diff
@@ -156,6 +79,99 @@ def unregister(self):
...
+
+        Args:
+            object_refs (List[ray.ObjectRef]): The object references
+                for the output (leaf) Ray tasks of the task graph.
+            dsk: The Dask graph.
+        """
+        pass
+
+    def _ray_finish(self, result):
+        """Run after all Ray tasks have finished executing and the final
```
Comment (angelinalg) [inline python/ray/util/dask/callbacks.py:165]:
```suggestion
        """Run after Ray finishes executing all Ray tasks and returns the final
```

File: python/ray/util/dask/callbacks.py [hunk 17]
```diff
@@ -156,6 +79,99 @@ def unregister(self):
...
+        result has been returned.
```
Comment (angelinalg) [inline python/ray/util/dask/callbacks.py:166]:
```suggestion
        result.
```

**Review by angelinalg — APPROVED** [review]
Just some style nits. Please correct if I've mangled the accuracy.

**Review by c21 — APPROVED** [review]

**Comment by peytondmurray** [comment]
@aslonnie If this looks okay to you, can we merge this?

**Comment by aslonnie** [comment]
@c21 or @angelinalg  should merge this

**Pull request merged**